### DISCUSSION TITLE
Any ideas on how to build a project that relies on flex and bison for Windows?

### DISCUSSION THREAD

**Discussion by Nemoumbra** [msys2/MSYS2-packages#4010] [Q&A]
The project I'm referring to is my CompilationTheory repo (for your convenience, that's the [link](https://github.com/Nemoumbra/CompilationTheory/tree/main)). It was originally created when I was working from my Linux machine and it works extremely great there. But it would be a shame if I couldn't deliver the tools that rely on flex and bison to the Windows users (including myself!).
Well, I've got a MSYS2 terminal installed with a bunch of MinGW64 standard apps.
![image](https://github.com/msys2/MSYS2-packages/assets/62447396/75695cf6-6a76-4e37-bf96-f729a6275547)
I assumed the flex and bison may be kind of non-typical tools so they could not have been ported yet to Windows... But then I checked and saw that it's not true. So to prepare I did `pacman -S flex`, `pacman -S bison`.
Then I've cloned my project with CLion and tried to build it. Well, the `flex.exe` and `bison.exe` apps seem to be functioning, but the IDE failed to compile my code because of a single include...
The file `FlexLexer.h` ended up in `k:\msys64\usr\include`. The toolchain that CLion calls "MinGW" is located in `k:\msys64\mingw64`.

This makes me wonder if I'm doing something wrong. Is there an example of how one can build something that uses flex and bison for Windows? There's an issue https://github.com/msys2/MSYS2-packages/issues/1911 where people were discussing this missing file. I agree that it's not a bug on MSYS2's side, but that doesn't concern me. Of course, I don't mean that you guys owe me something. I just want a friendly advice from people on how to achieve the expected result. Maybe there's another implementation of flex&bison for Windows that you'd recommend? If not, then what do you think I should do?

**Comment by sskras** [comment]
MSYS2 brings actually two worlds to you:

1. MSYS2-runtime (closely related to Cygwin, an originator of the codebase, suitable for building POSIX code)
2. MinGW environment (suitable for building WinAPI code)

It looks to me like CLion tries to use the _second_ one while you started going with the _first_ one. I guess that's the cause.

A significant part of users don't get that distinction initially.

And while I am not sure about the spectrum of Flex & Bison implementations on Windows (let alone the ways CLion can be set up on Windows), I would:

3. Try building the project using purely CLI in MSYS2 (and outside of MINGW64) – probably by running `cmake` commands.
4. Try doing that on Cygwin where you have no MinGW that could interfere.
5. Probably try on Midipix (a POSIX-on-Windows alternative to MSYS2 and Cygwin) – but it's still pre-alpha, so YMMV.

PS. Have you tried to build you project from the command line on Linux?
EDIT: I would be interested in the working command line.

**Reply by Nemoumbra** [reply]
Hmm... Now that you mention it, I've got a bunch of MSYS2 terminals (or runtimes, which I suppose would be a more accurate term):
![image](https://github.com/msys2/MSYS2-packages/assets/62447396/509e2522-d497-4b8e-9fd5-f07d56b2dff3)


For this whole time I've been using the MinGW64 one. The MSYS2 MSYS is so raw it doesn't even have cmake installed... OOPS.
```
 MSYS /k/Shared_storage/CLion/CompilationTheory/build-test
$ cmake ..
-bash: cmake: command not found
```
When it comes to CLion... I don't even know where it was supposed to look for the MSYS runtime.
![image](https://github.com/msys2/MSYS2-packages/assets/62447396/d678598d-fab8-4089-ab8b-c0c89c111181) (don't mind the test folder).
Next, I'm not sure what Cygwin is in my environment... Is it a standalone package with libraries, headers and a terminal? I also don't have the Midipix installed. I guess, not much a problem, but I'd like to deal with weird stuff as it appears.

PS: I didn't use CLion a year ago. It was VSCode for editing code + launching my Cmake files. The old-fashioned `cd build`, `cd build`, `cmake ..`, `make`. I certainly remember doing this from both the VS Code terminal and the GUI.

**Reply by sskras** [reply]
> I've got a bunch of MSYS2 terminals (or runtimes, which I suppose would be a more accurate term):
> ![image](https://user-images.githubusercontent.com/62447396/265549135-509e2522-d497-4b8e-9fd5-f07d56b2dff3.png)

True, terminals for MSYS2 are a bit different area:
https://www.msys2.org/docs/terminals/

And the guys I talk about are just called environments in the docs:
https://www.msys2.org/docs/environments/

> For this whole time I've been using the MinGW64 one. The MSYS2 MSYS is so raw it doesn't even have cmake installed... OOPS.
> 
> ```
>  MSYS /k/Shared_storage/CLion/CompilationTheory/build-test
> $ cmake ..
> -bash: cmake: command not found
> ```

That's because you've got installed the `mingw-w64-x86_64-cmake` package (for MINGW64 env), but not `cmake` (for MSYS2 env). Eg. I have installed both on my setup:

![image](https://github.com/msys2/MSYS2-packages/assets/7887758/daf38905-987d-4f49-8e53-7e8df750a9d7)

I suggest doing this from whichever env of MSYS2, and you should be fine:
```
$ pacman -S cmake
```

> Next, I'm not sure what Cygwin is in my environment... Is it a standalone package with libraries, headers and a terminal?

Something like that, only that it's a separate software bundle (independent from MSYS2) that has its' own installer and is set up into different directory. Actually it has two installers: a 32-bit and a 64-bit one.

It comes with the same terminal emulator MinTTY and with packages (but no CLI package manager) that bring you libs, headers and executable tools: https://www.cygwin.com/faq.html#faq.what.what

![image](https://github.com/msys2/MSYS2-packages/assets/7887758/7fd64340-7081-4095-869f-ad37f04dae49)

Just like MSYS2 minus MINGW64:

![image](https://github.com/msys2/MSYS2-packages/assets/7887758/ad410970-4e62-4154-88ee-7c8e2ccb1368)

Cygwin feels clumsy to manage to me, so if you find out how to set up your MSYS2 to fit your needs, please ignore this suggestion of mine.

**Reply by Nemoumbra** [reply]
I've got some news. Right after writing my last message here I've decided to try doing `pacman -S cmake`... And also `make`, `ninja`, `ccache`, stuff like that. Then I tried building the project and succeeded! So yeah, the MSYS2 MSYS runtime worked! Too bad there's nothing for MinGW... Maybe it won't be hard to port flex and bison to MSYS2 MinGW?
Unfortunately, it's not the whole story. It looks like CLion has got support for the following building toolchains:
![image](https://github.com/msys2/MSYS2-packages/assets/62447396/14146f59-0622-42c4-b6d9-921f108c0ca6)

If a dev wants to use a custom toolchain, they need to select "System" and then configure everything themselves. This can be helpful when you're cross-compiling something... But this time I've decided to use it to make me a MSYS toolchain. Have a look at this:
![image](https://github.com/msys2/MSYS2-packages/assets/62447396/bf86bd51-5e0a-4880-aa36-7a42053ba821)

Well, I've run Cmake from CLion and it executed successfully, then I have built it with CLion! Looks like everything's fine now, but there's a slight problem. The static code analyzer that the IDE provides is unable to locate the standard C++ library headers!
![image](https://github.com/msys2/MSYS2-packages/assets/62447396/57f94784-6bcb-4098-98e4-127c24ab7e92)
It doesn't have any troubles with finding `unistd.h` in `K:\msys64\usr\inlcude\unistd.h`. The actual path for `iostream` is `k:\msys64\usr\lib\gcc\x86_64-pc-msys\11.3.0\include\c++\iostream` so I have to figure out how to add the `k:\msys64\usr\lib\gcc\x86_64-pc-msys\11.3.0\include` path to a list of dirs for CLion to consider. Idk how that's supposed to be done, just wanted to let you know how I'm progressing (and for future generations of people who will try to use MSYS in CLion).
